### ISSUE TITLE
[HOLD for payment 2024-07-22] LOW: Enable Concurrent React Features

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#33531]
### Problem

There are a number of performance issues that we have long-term aspirations to solve via the [React concurrency features introduced in React 18+](https://github.com/reactwg/react-18/discussions/46). This was a big portion of the motivation to enable the new React Native architecture.

While we are still working on the New Arch on iOS and Android, we already have the ability to enable React concurrent mode on web. However, during the [RNW 0.19 upgrade](https://github.com/Expensify/App/issues/16660), it was found that some aspects of our codebase, particularly `withOnyx`, are not compatible with concurrent mode, so we turned it off.

### Solution
Let's investigate what it will take to re-enable React Concurrent Mode and start working on that in parallel to the New Arch upgrade, which is well over a year in the making and should land _hopefully_ by the end of January.

<details><summary>Issue Owner</summary>Current Issue Owner: @muttmuure</details>

**Comment by roryabraham** [comment]
Reassigning to @muttmuure for tracking as part of the NewDot Performance project

**Comment by roryabraham** [comment]
Put out a call for volunteers: https://expensify.slack.com/archives/C05LX9D6E07/p1703292300758289

**Comment by roryabraham** [comment]
Still looking for a volunteer, but I assume the people I tagged in slack may be partially OOO. This will remain on HOLD for this week, so I'll bump next week if I haven't heard anything

**Comment by roryabraham** [comment]
On HOLD for New Arch

**Comment by roryabraham** [comment]
On HOLD for [New Arch](https://github.com/Expensify/App/issues/8503)

**Comment by roryabraham** [comment]
Still on HOLD

**Comment by roryabraham** [comment]
Though, we could try to get a volunteer to start looking into enabling this on web to get a headstart?

**Comment by Szymon20000** [comment]
Ready for assignment :)

**Comment by roryabraham** [comment]
For now, the only platform that's ready for this is web, so you can focus your efforts there to start (while we're working on New Arch)

**Comment by Szymon20000** [comment]
Hello @roryabraham !
I spent around 3h today on investigating it and wanted to share my thoughts.
I would split the work here into 2 parts. 
The first one is making sure onyx is ready for the concurrent rendering and the second one is that all of our components in the app don't have any side effects that could break concurrent rendering. 

When it comes to onyx one I see 2 potential issues:
-  The first one is that we have optimistic reads from cache in withOnyx component. Because concurrent rendering can (and most likely will) split the rendering into multiple parts it's possible that some of the components will read newer state than the others. We probably will need to use `useSyncExternalStore` here (Maybe by wrapping withOnyx in functional component). We already have `allowStaleData` option so maybe reading inconsistent state is fine in this case.

- Second one is of course sending updates to withOnyx components. What can happen now is that we will call setState on the first withOnyx listener but the rest will be batched later so again we ends up inconsistent state as some components already know the newer value. I remember that someone was already working on adding Jotai to onyx. Jotai already has support for react concurrent rendering and would solve that problem. What's the current state of it? If we don't plan to use Jotai in onyx we will need to use useSyncExternalStore hook.

For the second part (so making sure all components don't have any dangerous side effect) we can use StrictMode component from React (https://react.dev/reference/react/StrictMode#strictmode) and just traverse the whole app too everything works properly and we don't have any warning on the console. I already added that locally and the report screen stopped to work.

**Comment by roryabraham** [comment]
> The first one is that we have optimistic reads from cache in withOnyx component ... We probably will need to use useSyncExternalStore here

Fortunately we have an [issue to implement a `useOnyx` hook](https://github.com/Expensify/App/issues/34339) and it's happening! The current proposal is to use the `useSynxExternalStore` hook 👍🏼 

> I remember that someone was already working on adding Jotai to onyx ... What's the current state of it? If we don't plan to use Jotai in onyx we will need to use useSyncExternalStore hook.

This is the first I've heard of adding Jotai to Onyx, but as I just mentioned we'll be setting up `useSyncExternalStore` soon. Maybe we'll have to update `withOnyx` to internally use `useOnyx`?

> For the second part (so making sure all components don't have any dangerous side effect) we can use StrictMode component from React (https://react.dev/reference/react/StrictMode#strictmode) and just traverse the whole app too everything works properly and we don't have any warning on the console

👍🏼 to enabling strict mode on dev. Maybe with a config-level flag to easily enable/disable it so we can turn it off and on easily.

> I already added that locally and the report screen stopped to work

Great 🙃 

So it sounds like this is the best thing to investigate next? Or do you think you'll be blocked on Onyx needing to use `useSyncExternalStore`?

**Comment by Szymon20000** [comment]
I spent another 2.5h to check what was the problem with report screen. 
It turned out that react navigation is not updating the params because it things the component has been unmounted. 
![image](https://github.com/Expensify/App/assets/12784455/b4ca8c29-0069-464d-be5b-c03f354b0951)
![image](https://github.com/Expensify/App/assets/12784455/de395c75-a0d8-4c03-9f3b-eab747005d91)
![image](https://github.com/Expensify/App/assets/12784455/ec3afb2c-a994-4cd9-8238-70b3dc282215)
It was recommended in the docs (https://es.legacy.reactjs.org/blog/2022/03/29/react-v18.html) to turn on the StrictMode but looks like it checks cases that don't really occur on production with concurrent mode. 
That's true that useMemo can be called multiple times with the same deps in concurrent mode but that shouldn't be true for useEffect. The StrictMode does checkes other important things. For instance Gesture Handler seems to use findNodeHandle that is deprecated in react 18. But for now I would wait for the new useOnyx hook that is based on useSyncExternalStore before we test the app further. 
In the meantime I will double check if running useEffect twice (with empty dips array) is actually possible on production wit the concurrent mode on.

**Comment by roryabraham** [comment]
Thanks @Szymon20000!

> For instance Gesture Handler seems to use findNodeHandle that is deprecated in react 18

Is there any upstream issue to clean this up? If not, can we create one?

**Comment by Szymon20000** [comment]
@roryabraham Not yet I think but I will open one. Looks like we can open another one in react navigation as that library doesn't work with StrictMode. 

I asked the question about StrictMode in react repo https://github.com/facebook/react/issues/28347 and they replied that calling useEffect twice can happen now with fast Refresh and will happen in the future on production with offscreen API. Some other hooks can be already be called twice on production with preserving the state. 

I think for now we can just patch react-navigation and open them a pr with that simple change that will make their problematic useEffect symmetrical. Then use it in our app to catch side effects that won't work with concurrent rendering. If at certain point we will see that one of our deps doesn't work well with strictMode we can just remove it and continue testing the app with concurrent mode manually. The only issue is that a lot of concurrent cases are rare and hard to catch without the StrictMode component.

**Comment by roryabraham** [comment]
Thanks for plugging away at this. Would love to see upstream issues created with minimal reproductions in react-navigation and RNGH. That would be great next steps and either you can work on fixing them if you choose or we can ask the maintainers to get involved. Those sound like steps we'd want to take before we add new patches imo

**Comment by roryabraham** [comment]
any updates in the last week?

**Comment by Szymon20000** [comment]
Hello @roryabraham !
We are waiting for useOnyx hook to avoid false-positive StrictMode issues. In the meantime @perunt will open issues in react-navigation as well as in RN-gesture-handler.

**Comment by roryabraham** [comment]
great, going to place this on HOLD for https://github.com/Expensify/react-native-onyx/pull/462

**Comment by muttmuure** [comment]
@roryabraham can we take this off hold now that the linked PR is merged?

**Comment by muttmuure** [comment]
Are we going to include this in the Onyx bump? https://github.com/Expensify/App/pull/38114

**Comment by roryabraham** [comment]
yeah, let's move this to be on HOLD for https://github.com/Expensify/App/pull/38114

**Comment by roryabraham** [comment]
`useOnyx` has landed in E/App - taking this off HOLD

**Comment by roryabraham** [comment]
Sounds like @chrispader is going to be taking this one over soon: https://expensify.slack.com/archives/C035J5C9FAP/p1713477420910129?thread_ts=1713455927.317579&cid=C035J5C9FAP

**Comment by roryabraham** [comment]
this is on the backlog - it does promise some good performance boosts but isn't particularly urgent for the roadmap

**Comment by roryabraham** [comment]
The feature I'm personally most excited for is [useTransition](https://react.dev/reference/react/useTransition) - a hook that allows you to perform state updates like `useState`, without blocking the UI while those updates are happening.

**Comment by muttmuure** [comment]
Not overdue

**Comment by mountiny** [comment]
@Szymon20000 Thanks for the exploration on this one so far.

Now as we are in a better place with the New Architecture, we would like to switch gears on this one and keep up the pressure so we can use React 18 and its relevant features. We would like to keep the limited resources of Margelo focused on the comment linking/ pagination which is high priority, so we are going to move this task over to different agency with goal of getting daily updates on next steps here.

**Comment by chrispader** [comment]
Hey @mountiny @roryabraham !

That sounds good. We've been struggling to make time for this feature so this makes sense! We'd want to push this as fast as possible now that New Arch enabled! 🙌🏼

We're actually talking to @WoLewicki at App.js Conf about pushing this forward right now, so handing this over is gonna be smooth...

**Comment by mountiny** [comment]
Thanks for understanding, we agree that it would be great to make a steady daily progress on this task now.

**Comment by mountiny** [comment]
Linking also an issue to research the react compiler which is going to be related https://github.com/Expensify/App/issues/42211

**Comment by Kicu** [comment]
hey I will be taking over and working on this :)

**Comment by Kicu** [comment]
I started to look at this today (only 3-4hrs):

**Good:**
 - if we turn on the concurrent mode, web app builds and runs and there are no obvious problems (although might be smaller ones, will continue to test)

**Bad**
 - while app works on concurrent mode, adding `<StrictMode>` breaks part of the app, and according to what I see the culprit is react-navigation, which doesn't work with StrictMode. So basically consistent with this previous finding: https://github.com/Expensify/App/issues/33531#issuecomment-1951879159
 Relevant issue in `react-navigation` about strict mode: https://github.com/react-navigation/react-navigation/issues/10988
 - I suspect just like described above the `withOnyx` hook might still be problematic but I need to investigate further - when I wrapped a new `SearchPage` screen in `StrictMode` it worked - this is a pretty fresh screen that uses only `useOnyx` hook - so this change is definitely promising

Technically we don't _NEED_ `StrictMode` for the app to work. It's a dev mechanism to find potential bugs, but I think its unsafe to move forward without it.
This is all preliminary and I will provide more updates as I dig further into this.

**Comment by Kicu** [comment]
Today a short update from me: most of the day I had to take care of conflicts on my PR related to Search which had a high priority as Search is on a deadline.
Next week I will be back with more updates.

That being said in my opinion the next biggest step to move this forward is to try and migrate the app onto `useOnyx` hook, as this hook is using [useSyncExternalStore](https://react.dev/reference/react/useSyncExternalStore) which is the only safe way to read data in concurrent React.
I will start research how easy it will be to do this, most likely via reimplementing `withOnyx` hoc using `useOnyx` hook.

**Comment by roryabraham** [comment]
> I will start research how easy it will be to do this, most likely via reimplementing withOnyx hoc using useOnyx hook.

This seems doable in the short-term. `withOnyx` returns `null` if data is still loading from Onyx, but `useOnyx` returns an `isLoading` state. We could use `useOnyx` internally in `withOnyx`, and use that `isLoading` state and to return `null` early in `withOnyx`

I also might like to see us invest more in the "real" fix which would be to migrate our E/App components from `withOnyx` to `useOnyx`. That hasn't been a priority in the past because (outside of it being hard to use with TypeScript), there hasn't been a real problem with `withOnyx`. Maybe we could consider doing what we did for the [light theme migration](https://github.com/Expensify/App/pull/30044/files#diff-08db062cf8791ec09dd2da983704e69225709821fe9b57e65c4f6fd62f51d51e), and create a codemod / babel script that automatically converts components from using `withOnyx` to `useOnyx`? That could end up actually being pretty easy to do technically - the challenge is that we don't know when the render-blocking behavior of `withOnyx` matters or not...

**Comment by mountiny** [comment]
> withOnyx returns null if data is still loading from Onyx, but useOnyx returns an isLoading state. We could use useOnyx internally in withOnyx, and use that isLoading state and to return null early in withOnyx

@chrispader is now working on updating this, and Onyx will return undefined if the key is not available yet. This will definitely require some changes in the app https://github.com/Expensify/react-native-onyx/pull/554 

I agree with rory though that we should maybe invest more effort into migrating over to useOnyx from withOnyx

**Comment by Kicu** [comment]
Yeah I also agree that in the long run fully switching to `useOnyx` is the best way and definitely worth it. The problem with `withOnyx` is that it _might_ affect how the app works and it might _not_ - you'll never know. It might work right now but break in a month after some new code is merged.

However I assumed that it would take too long to rewrite everything to the hook so I thought that rewriting `withOnyx` might be faster. Hard for me to guess how difficult the codemod might be - but I think non trivial since types of all the props will change and those would have to be updated manually perhaps? - with HOC you wrap component so it has a combination of props, hooks you use inside so the values from onyx no longer enter via props. 

Also I'm unsure about: https://github.com/Expensify/react-native-onyx/pull/554 - we would have to migrate web onto this right? since not returning null sounds like BC - do I understand correctly?

What do you guys think I should focus now? I think the main difficulty right now is discerning what app problems come from `react-navigation` and what come from usage of `withOnyx`.

**Comment by Kicu** [comment]
small update from me:
For the last 2 days I didn't spend a lot of time on this task, because I was finishing up sorting in Search and I also tried to dig up my old PR here: https://github.com/Expensify/App/pull/39072 - which is way overdue.

But now I'm almost done and tomorrow I will focus much more on concurrent React.
I have one finding, which I already discussed with @blazejkustra and wanted to discuss here:

- The new `useOnyx` hook allows us to connect to 1 onyx value/collection (see: https://github.com/Expensify/react-native-onyx/blob/main/lib/useOnyx.ts#L78)
- However `withOnyx` hoc allowed us to subscribe to multiple keys which we often do (case in point: https://github.com/Expensify/App/blob/main/src/pages/home/ReportScreen.tsx#L734-L765)

Because React hooks cannot be added/removed dynamically, actually rewriting internal implementation of `withOnyx` to use `useOnyx` will be non trivial and flaky. You can't just call `useOnyx` in a loop within the component. I believe it technically _can_ be done, but it's not code that we would want on production anyway. Perhaps only for local testing.

So the cleaner way forward would be to rewrite app fully to `useOnyx`.
Because each case of 1 call to `withOnyx` will map to multiple usages of `useOnyx` hook, I think it will be safer to do it in batches manually by developers. Codemod might help, but IMO this migration will require a lot of extra manual tweaking.

My plan for next 1-2 days is to dig into the App and try to discern how much of possible problems comes from `react-navigation` vs from `withOnyx` usage. The migration might take a while, so first I want to make sure it's worth it - perhaps concurrent features would still work even on our current code 🤔 🤷 

Meanwhile please tell me if you want me to focus on something more specific or if you have any more questions.
@roryabraham @mountiny

**Comment by mountiny** [comment]
@Kicu thanks for the summary, I think your next steps sound good and we would eventually want to migrate to useOnyx but it will take a bit indeed.

**Comment by Kicu** [comment]
I spent some hours in the past 2 days digging into this.

### new findings
I have found a root cause of a few issues and that is actually something new - animations from `react-native-reanimated`, which seems to misbehave when inside `<StrictMode>`. Or at least some of our use cases break. This is something that I will be focusing on next week and I will definitely ask for help OS devs from SWM how maintain this library.
This is important because some of the animations result in big problems for the app. For example we are animating `ReportScreen` when switching chats on desktop from `opacity` `0` to `1` and if this animation glitches, opacity stays as zero, thus the user is not able to see the report. 

### status
However once I understood this is the problem and disabled/commented out some of the animations the whole App seems to behave mostly correctly, if a bit glitchy on the visual end.

The good news is that I now believe that the usage of `withOnyx` is actually **not a big problem** right now - even though in future we should definitely migrate to `useOnyx` sooner rather than later.
Also `react-navigation` seems to behave correctly after applying a small patch, however that is one more thing to investigate next week. Not sure yet whether the patch is sufficient and whether it doesn't break other things inside navigation - because it's quick&dirty right now.

#### glitching animation of FAB
https://github.com/Expensify/App/assets/3929868/657a5d1c-2db0-4fd8-a2bd-e7a3eaf4187e

#### glitch then ReportScreen stays at opacity 0
<img width="1100" alt="Screenshot 2024-06-07 at 16 38 31" src="https://github.com/Expensify/App/assets/3929868/7d7106cd-fd9d-48e5-99e8-1a4df47f31ae">


More updates to come next week.

**Comment by roryabraham** [comment]
Yesterday I announced the [official deprecation of `withOnyx`](https://expensify.slack.com/archives/C01GTK53T8Q/p1717695182186119), so at least we should not allow any _new_ uses.

**Comment by quinthar** [comment]
Thanks for these updates!

**Comment by Kicu** [comment]
Small update from me:

### new findings
With the help of a developer from `reanimated` OS team I was able to patch parts of reanimated lib that solved a lot of our problems. They will push this fix soon to upstream, so we should be able to update.
There is 1 animation bug right now which I temp-solved by commenting out this animation. (It's the exit fade out animation for mentions list): https://github.com/Expensify/App/pull/42592/files#diff-82ae7ae5193fe8f0f5d3cc064489bd3fdb78b2fc9738a7649a510fbf5864fe12R89

Now I'm fighting a very complicated bug that is 90% related to react navigation, where if use enters on root url `/` then they are not redirected to last visited reportID - which they should be.
Because `StrictMode` renders every component twice and then runs `useEffect` twice - these kinds of small bugs are hard to understand.

I'm giving myself time till tomorrow to try and understand this, and if not I will somehow move forward.

I expect to come back with bigger update about next steps in 1st half of next week.

**Comment by Kicu** [comment]
For the past 3 days I've been mostly trying to understand how `react-navigation` breaks on StrictMode, because that was our biggest blocker.

Right now in `StrictMode` entering initial screen from deeplinks does not work, and you always get redirected to Home.
I have now understood mostly what happens (with great help from @adamgrzybowski 🎉 ) and I created a pure reproduction of this error here: https://github.com/Kicu/react-navigation-strictmode
Next steps is to show it to maintainers of the library and ask them for help or fix.
For us temporarily I can do some quick-and-dirty patch that will make the app work.

Other than `react-navigation` I have 2 patches for `reanimated` that should also make the lib work with `StrictMode`.
I think I need 1-2 more days to have a version of this that is ready to be tested by more people.

As usual more updates will come in a few days 🙃

**Comment by muttmuure** [comment]
Will concurrent mode improve composer behavior? I've noticed if you type while heavy commands are pending (like OpenApp) the focus is erratic and jumps back to the start position. 

If you compare with other apps like Slack, focus is much more stable

**Comment by Kicu** [comment]
@muttmuure 
Maybe it could help, however I don't know `composer` related code well enough to say that with certainty. Soon we will be able to test the app and we will see what happens.

However this task is mostly about web version of the App, since on native I believe the New Architecture should already be running in concurrent mode so if you notice composer slowness there we might need to dig deeper into this.
But we can keep exploring this in future

**Comment by mountiny** [comment]
Yep I agree its best now to ask QA to test this, there are some other adhoc build testing going on now so it might betake a week to complete

**Comment by roryabraham** [comment]
Great, the PR has just been merged. Soon, we'll want to follow-up with an investigation in #newdot-quality about how features such as Suspense and `useTransition` can be used to improve the responsiveness of the app

**Comment by muttmuure** [comment]
Upwork job got mangled

**Comment by muttmuure** [comment]
$250 for @rayane-djouah

**Comment by muttmuure** [comment]
Offer sent

**Comment by muttmuure** [comment]
Paid
